### PR TITLE
added AMSgrad optimizer to Adam and SparseAdam

### PR DESCRIPTION
**Pull request by kashif**
- [x] need to look at SparseAMSGrad version

### CODE REVIEW

**Comment by kashif** [comment]
ok ready for a review!

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
thanks @kashif !

File: torch/optim/sparse_adam.py [hunk 1]
```diff
@@ -83,7 +95,14 @@ def make_sparse(values):
 
                 # Dense addition again is intended, avoiding another _sparse_mask
                 numer = exp_avg_update_values.add_(old_exp_avg_values)
-                denom = exp_avg_sq_update_values.add_(old_exp_avg_sq_values).sqrt_().add_(group['eps'])
+                exp_avg_sq_update_values.add_(old_exp_avg_sq_values)
+                if amsgrad:
+                    torch.max(old_max_exp_avg_sq_values, exp_avg_sq_update_values, out=old_max_exp_avg_sq_values)
+                    denom = old_max_exp_avg_sq_values.sqrt_().add_(group['eps'])
+                    max_exp_avg_sq = make_sparse(old_max_exp_avg_sq_values)
```
Comment (apaszke) [inline torch/optim/sparse_adam.py]: You aren't updating the saved `max_exp_avg_sq` in here. You're only assigning it to a local. Also, you don't want to retain max as a sparse tensor, but as a dense tensors that you'll be indexing out with non-zero grads at this step, and only scattering updates for those entries.

**Review by apaszke — COMMENTED** [review]
I think there's a bug in SparseAdam now

File: torch/optim/sparse_adam.py (hunk 1 above)
Comment (kashif) [reply torch/optim/sparse_adam.py]: thanks @apaszke! I see let me try to fix it...

**Pull request merged**